### PR TITLE
Update Demo to latest Comet Starter

### PR DESCRIPTION
**Pull request by piotrgulbis**
Update Demo to latest Comet Starter, keeping everything that was in Demo if possible.

### CODE REVIEW

File: demo/api/src/config/config.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export function createConfig(processEnv: NodeJS.ProcessEnv) {
         envVars.AZURE_OPEN_AI_CONTENT_GENERATION_DEPLOYMENT_ID
     ) {
         contentGeneration = {
-            apiKey: envVars.AZURE_OPEN_AI_CONTENT_GENERATION_API_KEY,
+            apiKey: envVars.AZURE_OPEN_AI_CONTENT_GENERATION_API_KEY ?? "",
```
Comment (johnnyomair) [inline demo/api/src/config/config.ts:22]: This shouldn't be necessary 🤔

File: demo/api/src/db/migrations/Migration20220127092535.ts
Comment (johnnyomair) [inline demo/api/src/db/migrations/Migration20220127092535.ts:1]: Please don't alter existing migrations, as they will not be executed again. Also, the changes made here a done in a later migration: https://github.com/vivid-planet/comet/blob/main/demo/api/src/db/migrations/Migration20220223131155.ts.

File: demo/api/src/db/migrations/Migration20220127092750.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ import { Migration } from "@mikro-orm/migrations";
 export class Migration20220127092750 extends Migration {
     async up(): Promise<void> {
         this.addSql(
-            'create table "Page" ("id" uuid not null, "content" json not null, "seo" json not null, "createdAt" timestamp with time zone not null, "updatedAt" timestamp with time zone not null);',
+            'create table "Page" ("id" uuid not null, "content" json not null, "stage" json not null, "seo" json not null, "createdAt" timestamp with time zone not null, "updatedAt" timestamp with time zone not null);',
```
Comment (johnnyomair) [inline demo/api/src/db/migrations/Migration20220127092750.ts:6]:
```suggestion
            'create table "Page" ("id" uuid not null, "content" json not null, "seo" json not null, "createdAt" timestamp with time zone not null, "updatedAt" timestamp with time zone not null);',
```

This is done in the migration below.

File: demo/api/src/footer/generated/footer.resolver.ts [hunk 3]
```diff
@@ -5,13 +5,13 @@ import { InjectRepository } from "@mikro-orm/nestjs";
...
 
+import { FooterScope } from "../dto/footer-scope";
 import { Footer } from "../entities/footer.entity";
-import { FooterContentScope } from "../entities/footer-content-scope.entity";
 import { FooterInput } from "./dto/footer.input";
 import { FootersService } from "./footers.service";
 
 @Resolver(() => Footer)
-@RequiredPermission(["pageTree"])
+@RequiredPermission(["footers"])
```
Comment (johnnyomair) [inline demo/api/src/footer/generated/footer.resolver.ts:14]: This should still be pageTree

File: demo/site/src/documents/pages/blocks/FooterContentBlock.tsx
Comment (johnnyomair) [inline demo/site/src/documents/pages/blocks/FooterContentBlock.tsx:1]: This should be in src/layout/footer/blocks

File: demo/site/src/util/graphQLClient.ts
Comment (johnnyomair) [inline demo/site/src/util/graphQLClient.ts:1]: Please revert the changes made here. Demo is newer than Starter.

**Comment by johnnyomair** [comment]
Some things that could be made as well
1. Update the folder structure to match the Starter
2. Remove outdated blocks from Demo, e.g., HeadlineBlock

**Comment by piotrgulbis** [comment]
> Some things that could be made as well
> 
> 1. Update the folder structure to match the Starter
> 2. Remove outdated blocks from Demo, e.g., HeadlineBlock

I was not sure here what to remove and what to keep... For example I removed the LinkList as indicated by your comment but we should keep it, so there is no way I can know which to keep and which to throw away.

File: demo/admin/src/pages/blocks/PageContentBlock.tsx
[the remark this answers is not in the document]
Comment (piotrgulbis) [reply demo/admin/src/pages/blocks/PageContentBlock.tsx:1]: Older PageContent removed.

File: demo/admin/src/pages/blocks/SeoBlock.tsx
[the remark this answers is not in the document]
Comment (piotrgulbis) [reply demo/admin/src/pages/blocks/SeoBlock.tsx:1]: Removed older SeoBlock

File: demo/api/src/db/migrations/Migration20220127092535.ts
Comment (piotrgulbis) [reply demo/api/src/db/migrations/Migration20220127092535.ts:1]: Removed edits from existing migration files

File: demo/site/src/documents/pages/blocks/FooterContentBlock.tsx
Comment (piotrgulbis) [reply demo/site/src/documents/pages/blocks/FooterContentBlock.tsx:1]: Moved to layout/footer/blocks

File: demo/api/src/footer/generated/footer.resolver.ts (hunk 3 above)
Comment (piotrgulbis) [reply demo/api/src/footer/generated/footer.resolver.ts:14]: This is an autogenerated file, need to find where this gets generated from.

File: demo/admin/src/common/apollo/createApolloClient.ts [hunk 4]
```diff
@@ -6,7 +6,6 @@ import fragmentTypes from "@src/fragmentTypes.json";
 export const createApolloClient = (apiUrl: string) => {
     const httpLink = new HttpLink({
         uri: `${apiUrl}/graphql`,
-        credentials: "include",
```
[the remark this answers is not in the document]
Comment (piotrgulbis) [reply demo/admin/src/common/apollo/createApolloClient.ts]: Added in again

File: demo/admin/src/common/blocks/TextLinkBlock.tsx [hunk 5]
```diff
@@ -2,4 +2,4 @@ import { createTextLinkBlock } from "@comet/cms-admin";
 
 import { LinkBlock } from "./LinkBlock";
 
-export const TextLinkBlock = createTextLinkBlock({ link: LinkBlock, name: "DemoTextLink" });
+export const TextLinkBlock = createTextLinkBlock({ link: LinkBlock });
```
[the remark this answers is not in the document]
Comment (piotrgulbis) [reply demo/admin/src/common/blocks/TextLinkBlock.tsx:5]: For what exactly?

File: demo/site/src/util/graphQLClient.ts
Comment (piotrgulbis) [reply demo/site/src/util/graphQLClient.ts:1]: Done.

File: demo/api/src/footer/generated/footer.resolver.ts (hunk 3 above)
Comment (piotrgulbis) [reply demo/api/src/footer/generated/footer.resolver.ts:14]: This is also footers in Comet Starter... Would this maybe get generated from the Service Name?

For example news also has "news" as requiredPermission and products has "products"

File: demo/admin/src/App.tsx
Comment (johnnyomair) [inline demo/admin/src/App.tsx:1]: Please keep the removed providers and configuration options.

File: demo/admin/src/common/MasterMenu.tsx
Comment (johnnyomair) [inline demo/admin/src/common/MasterMenu.tsx:1]: Please keep the removed routes

File: demo/admin/src/common/blocks/LinkListBlockWithAdditionalFields.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { createListBlock } from "@comet/blocks-admin";
+import { TextLinkBlock } from "@src/common/blocks/TextLinkBlock";
+import { userGroupAdditionalItemFields } from "@src/userGroups/userGroupAdditionalItemFields";
+import { UserGroupChip } from "@src/userGroups/UserGroupChip";
+import { UserGroupContextMenuItem } from "@src/userGroups/UserGroupContextMenuItem";
+import { FormattedMessage } from "react-intl";
+
+export const LinkListBlockWithAdditionalFields = createListBlock({
```
Comment (johnnyomair) [inline demo/admin/src/common/blocks/LinkListBlockWithAdditionalFields.tsx:8]: I wouldn't add a new link list block for this, keep it in the original block instead.

File: demo/admin/src/dashboard/DashboardPage.tsx [hunk 7]
```diff
@@ -18,15 +18,11 @@ const Dashboard = () => {
                     "2x": backgroundImage2x,
                 }}
             />
-            <Toolbar scopeIndicator={<ContentScopeIndicator global />} />
             <MainContent>
                 <Grid container direction="row" spacing={4}>
-                    {isAllowed("pageTree") && <LatestContentUpdates />}
```
Comment (johnnyomair) [inline demo/admin/src/dashboard/DashboardPage.tsx:24]: This check should probably stay and also be added to Starter.

File: demo/admin/src/documents/pages/mainMenu/MainMenu.tsx
Comment (johnnyomair) [inline demo/admin/src/documents/pages/mainMenu/MainMenu.tsx:1]: This shouldn't be in src/documents/pages, but in src/mainMenu or src/projectSnips/mainMenu

File: demo/admin/src/news/blocks/NewsContentBlock.tsx [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
+import { HeadingBlock } from "@src/common/blocks/HeadingBlock";
 import { RichTextBlock } from "@src/common/blocks/RichTextBlock";
 import { TextImageBlock } from "@src/common/blocks/TextImageBlock";
 
 export const NewsContentBlock = createBlocksBlock({
     name: "NewsContentBlock",
     supportedBlocks: {
-        headline: HeadlineBlock,
-        richtext: RichTextBlock,
+        headline: HeadingBlock,
```
Comment (johnnyomair) [inline demo/admin/src/news/blocks/NewsContentBlock.tsx:10]: The key should probably be heading

File: demo/admin/src/pages/PageContentBlock.tsx [hunk 9]
```diff
@@ -1,51 +0,0 @@
...
-    },
-    additionalItemFields: {
-        ...userGroupAdditionalItemFields,
-    },
-    AdditionalItemContextMenuItems: ({ item, onChange, onMenuClose }) => {
-        return <UserGroupContextMenuItem item={item} onChange={onChange} onMenuClose={onMenuClose} />;
-    },
-    AdditionalItemContent: ({ item }) => {
-        return <UserGroupChip item={item} />;
-    },
```
Comment (johnnyomair) [inline demo/admin/src/pages/PageContentBlock.tsx:50]: Please add this to the "new" page content block

File: demo/admin/src/pages/blocks/LayoutBlock.tsx
Comment (johnnyomair) [inline demo/admin/src/pages/blocks/LayoutBlock.tsx:1]: Please don't remove this Block

File: demo/api/src/common/blocks/link-list.block.ts [hunk 10]
```diff
@@ -1,18 +1,5 @@
-import { BaseListBlockItemData, BaseListBlockItemInput, BlockField, createListBlock } from "@comet/blocks-api";
-import { UserGroup } from "@src/user-groups/user-group";
-import { IsEnum } from "class-validator";
+import { createListBlock } from "@comet/blocks-api";
 
 import { TextLinkBlock } from "./text-link.block";
 
-class ListBlockItemData extends BaseListBlockItemData(TextLinkBlock) {
```
Comment (johnnyomair) [inline demo/api/src/common/blocks/link-list.block.ts:7]: Please keep this

File: demo/api/src/pages/blocks/layout.block.ts
Comment (johnnyomair) [inline demo/api/src/pages/blocks/layout.block.ts:1]: Please keep this.

File: demo/site/src/blocks/CookieSafeYouTubeVideoBlock.tsx
Comment (johnnyomair) [inline demo/site/src/blocks/CookieSafeYouTubeVideoBlock.tsx:1]: We should keep this.

File: demo/site/src/blocks/LayoutBlock.tsx
Comment (johnnyomair) [inline demo/site/src/blocks/LayoutBlock.tsx:1]: Please keep this.

File: demo/site/src/common/helpers/GridRoot.tsx
Comment (johnnyomair) [inline demo/site/src/common/helpers/GridRoot.tsx:1]: We can probably remove this.

File: demo/site/src/documentTypes/Page.tsx
Comment (johnnyomair) [inline demo/site/src/documentTypes/Page.tsx:1]: Please move this to src/documents/pages/ to match the folder structure in Admin & API.

File: docs/docs/blocks/block-factories.mdx [hunk 11]
```diff
@@ -547,7 +547,7 @@ export function OptionalImageBlock({ data }: PropsWithData<OptionalImageBlockDat
 The RichTextBlock factory can be used to create a block with a rich text.
 It requires a link block to be set.
 
-import RichTextBlock from "./images/rich-text-block.png";
+import { RichTextBlock } from "./images/rich-text-block.png";
```
Comment (johnnyomair) [inline docs/docs/blocks/block-factories.mdx:550]: I doubt that the image will work when imported like this.

File: demo/api/src/footer/generated/footer.resolver.ts (hunk 3 above)
Comment (johnnyomair) [reply demo/api/src/footer/generated/footer.resolver.ts:14]: You can change the permission in the API Generator: https://github.com/vivid-planet/comet/blob/main/demo/api/src/footer/entities/footer.entity.ts#L13

File: demo/admin/src/common/blocks/TextLinkBlock.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (johnnyomair) [reply demo/admin/src/common/blocks/TextLinkBlock.tsx:5]: Changing the name of the block. But this is out of scope.

[the remark this answers is not in the document]
Comment (piotrgulbis) [reply demo/admin/src/common/blocks/TextLinkBlock.tsx:5]: Okay, maybe in followup.

File: demo/api/src/footer/generated/footer.resolver.ts (hunk 3 above)
Comment (piotrgulbis) [reply demo/api/src/footer/generated/footer.resolver.ts]: Nice, didn't know ;) Updated!

File: docs/docs/blocks/block-factories.mdx (hunk 11 above)
Comment (piotrgulbis) [reply docs/docs/blocks/block-factories.mdx:550]: Renamed too much.

File: demo/admin/src/common/blocks/LinkListBlockWithAdditionalFields.tsx (hunk 6 above)
Comment (piotrgulbis) [reply demo/admin/src/common/blocks/LinkListBlockWithAdditionalFields.tsx:8]: This causes a lot of other problems because if kept in the original block, the LinkList is used in the Footer and there is then a type mismatch because of the addition of userGroup field.

File: demo/site/src/common/helpers/GridRoot.tsx
Comment (piotrgulbis) [reply demo/site/src/common/helpers/GridRoot.tsx:1]: It is used in components/Breadcrumbs.tsx

Comment (piotrgulbis) [reply demo/site/src/common/helpers/GridRoot.tsx:1]: But I will move it into Breadcrumbs

File: demo/site/src/blocks/LayoutBlock.tsx
Comment (piotrgulbis) [reply demo/site/src/blocks/LayoutBlock.tsx:1]: Restored.

File: demo/site/src/blocks/CookieSafeYouTubeVideoBlock.tsx
Comment (piotrgulbis) [reply demo/site/src/blocks/CookieSafeYouTubeVideoBlock.tsx:1]: Restored

File: demo/admin/src/documents/pages/mainMenu/MainMenu.tsx
Comment (piotrgulbis) [reply demo/admin/src/documents/pages/mainMenu/MainMenu.tsx:1]: moved to src/mainMenu

File: demo/admin/src/common/blocks/LinkListBlockWithAdditionalFields.tsx (hunk 6 above)
Comment (johnnyomair) [reply demo/admin/src/common/blocks/LinkListBlockWithAdditionalFields.tsx:8]: Please add `@ts-expect-error` and TODOs instead. Otherwise we'll never fix this. You could even create a bug task.

File: demo/admin/src/App.tsx [hunk 12]
```diff
@@ -1,13 +1,10 @@
-import "@fontsource-variable/roboto-flex/full.css";
```
Comment (johnnyomair) [inline demo/admin/src/App.tsx:1]: Is it okay to remove this font?

File: demo/api/src/footer/entities/footer.entity.ts [hunk 13]
```diff
@@ -1,30 +1,33 @@
...
 
 import { FooterContentBlock } from "../blocks/footer-content.block";
-import { FooterContentScope } from "./footer-content-scope.entity";
+import { FooterScope } from "../dto/footer-scope";
 
 @Entity()
-@ObjectType()
+@ObjectType({
+    implements: () => [DocumentInterface],
+})
```
Comment (johnnyomair) [inline demo/api/src/footer/entities/footer.entity.ts:13]: The document interface was intentionally removed in https://github.com/vivid-planet/comet/pull/1906. We should remove it in Starter as well.

File: demo/api/src/footer/footer.module.ts [hunk 14]
```diff
@@ -1,14 +1,12 @@
...
 import { Module } from "@nestjs/common";
 
 import { Footer } from "./entities/footer.entity";
-import { FooterContentScope } from "./entities/footer-content-scope.entity";
 import { FooterResolver } from "./generated/footer.resolver";
 import { FootersService } from "./generated/footers.service";
 
 @Module({
-    imports: [MikroOrmModule.forFeature([Footer, FooterContentScope])],
-    providers: [FooterResolver, FootersService, DependenciesResolverFactory.create(Footer)],
```
Comment (johnnyomair) [inline demo/api/src/footer/footer.module.ts:12]: Please don't remove the dependencies

File: demo/site/src/common/blocks/TextLinkBlock.tsx [hunk 15]
```diff
@@ -0,0 +1,12 @@
+"use client";
+import { PropsWithData, withPreview } from "@comet/cms-site";
+import { TextLinkBlockData } from "@src/blocks.generated";
+
+import { LinkBlock } from "./LinkBlock";
+
+export const TextLinkBlock = withPreview(
+    ({ data: { link, text } }: PropsWithData<TextLinkBlockData>) => {
+        return <LinkBlock data={link}>{text}</LinkBlock>;
```
Comment (johnnyomair) [inline demo/site/src/common/blocks/TextLinkBlock.tsx:9]: The (now deleted) old text link block had a special handling for DAM file downloads to show the file size. Please restore this behavior.

File: demo/api/src/footer/entities/footer.entity.ts [hunk 16]
```diff
@@ -1,16 +1,13 @@
...
+import { Field, ID } from "@nestjs/graphql";
 import { v4 } from "uuid";
 
 import { FooterContentBlock } from "../blocks/footer-content.block";
 import { FooterScope } from "../dto/footer-scope";
 
 @Entity()
-@ObjectType({
-    implements: () => [DocumentInterface],
-})
```
Comment (johnnyomair) [inline demo/api/src/footer/entities/footer.entity.ts]: Please start the API to update the schema.gql

**Review by johnnyomair — APPROVED** [review]
